### ISSUE TITLE
Schannel issues "failed to get certificate location" when using alternate certstore

### ISSUE THREAD

**Issue by rlmoore0** [jeroen/curl#198]
This issue is related to multiple other issues, which I've reviewed and tried for various solution hints.  

I am using curl 4.0 on a Windows 7, service pack 1 machine.  R version 3.6.1.

I am in a corporate environment pulling data from a government system.  The government system is not reachable on the Internet.  The government creates all of the certificates being used, both client and server.  (The network uses dual authentication.)

(Note:  my apologies for any typos below.  The government-connected system is not connected to anything else, which means I have to retype the code and output below.)

**Running the code:** 

h <- curl::new_handle()
curl:: handle <-setopt(h, .list = list(
cainfo = "U:/Private/Certificates/Apache _Bundle_AllTrustedPartners.crt", sslcert = "U:/private/Certificates/rlmoor1.ds_complete.pem",
keypasswd = "mykeypass",
userpwd = "rlmoorl:myuserpwd",
verbose = 1))

temp <- curl::curl_fetch_memory("https://jira.ud.tools.gov/jira/rest/auth/1/session", h)

**Gives the following output:**

* Trying 8.175.43.168...
* TCP NODELAY set
* Connected to jira.ud.tools.gov (8.175.43.168) port 443 (#0)
* schannel: Failed to get certificate location for U :/private/Certificates/rlmoorl.dscomplete.pem
* Closing connection 0
* schannel: shutting down SSL/TLS connection with jira.ud.tools.gov port 443
Error in curl::curl_fetch_memory("https://jira.ud.tools.gov/jira/rest/auth/1/session", schannel: Failed to get certificate location for
U :/private/Certificates/rlrnoorl .ds_complete. pem

**Here's the curl version information:**

$version
[1] "7.64.1"
$ssl_version
[1] "(OpenSSL/1.1.1a) Schannel"
$libz_version
[1] "1.2.8"
$libssh_version
[1] "Iibssh2/1.8.2"
$libidn_version 
[1] NA
$ host
[1] "x86_64-w64-rningw32"
$protocols
[1] "dict" file" "ftp" "ftps" "gopher" "http" "https" "imap" "imaps" "Idap" "Idaps" "pop3"
[13] 11pop3s" "rtsp" "scp" "sftp" "smtp" "smtps" "telnet" "tftp"
$ipv6
[1] TRUE
$http2
[1] FALSE
$idn
[1] TRUE

**If I regress to curl 2.8.1, the code works.  (Returns a code 200 with appropriate content.)**

**Also, if I invoke git's curl implementation directly through R as follows:**

system2("C:/Program Files/Git/mingw64/bin/curl.exe", args = c(" --cacert U:/private/Certificates/Apache _Bundle_AllTrustedPartners.crt", " --cert
U:/private/Certificates/rlmoor1.ds_complete.pem", "--pass mykeypass", "--user rlmoor1:myuserpwd", "https://jira.ud.tools.gov/jira/rest/auth/1/session"), stdout = TRUE)

**Then I also get success.**

**For git's curl version, I have:**

system2("C:/Program Files/Git/mingw64/bin/curl.exe", args = c("--version"), stdout = TRUE)

[1] "curl 7.61.1 (x86_64-w64-mingw32) libcurl/7.61 .1 OpenSSL/1 .0.2p (WinSSL) zlib/1.2. 11 libidn2/2.0.5 nghttp2/1.330"
[2] "Release-Date: 2018-09-05"
[3] "Protocols: dict file ftp ftps gopher http https imap imaps Idap Idaps pop3 pop3s rtsp smtp smtps telnet tftp"
[4] "Features: AsynchDNS IDN IPv6 Largefile SSPI Kerberos SPNEGO NTLM SSL libz TLS-SRP HTTP2 H1TPS-proxy MuIt1SSL Metalink"

I **think** the problem is that certificates are not being stored in the Microsoft certstore accessed with schannel.  I know nothing is in my Microsoft certstore on my machine.  Why no certstore?  Because the government has created an alternate certstore that they control.  I can export my certificates from that certstore and convert my certs to PEM using OpenSSL, but only system administrators have the authorization to add certs to the Microsoft certstore and the government forbids that.  curl does not know about that alternate cert store.

I talked with a Python developer who was familiar with the problem.  He said the only workaround (for Python) was to make changes in ssl.c to direct it to either openssl or schannel.  I'm not sophisticated enough with the protocols to want to change the equivalent code in curl _and_ I don't care to become disconnected from the main development path for curl.  I could revert to curl 2.8.1, but that's also not great.

So . . . help, please.

**Comment by jeroen** [comment]
Currently we do not support `cainfo` on Windows because indeed because it [doesn't work well](https://github.com/curl/curl/issues/3711). Maybe that will change soonish but that is out of my control.

There are two solutions:

 1. Import your CA bundle into the Windows cert store.
 2. Disable cert validation in your R code.

If you can't modify the cert store, I think (2) is the only way forward for you. So you need to set `ssl_verifypeer` to FALSE in your call to `handle_setopt()` and then you should be able to make the connection.

**Comment by rlmoore0** [comment]
Thanks!--although I won't be able to test disabling cert validation until Monday, since it is on a work-only system that I won't have access to until then.

Why did this work under curl 2.8.1 but not now?  I know mandatory schannel was added after 2.8.1.

Unfortunately, loading anything into Windows cert store isn't viable, since doing so is forbidden by government policy.  (The government supplies an alternate cert store.)

Note that the certificate that is generating the error message is about **my** certificate, not the CA bundle.  The government's system does _dual_ authentication:  my cert proves my identity to the server, and the server certificate proves the server's identity to me.  I can disable the authentication of the server down to me, but I can't disable the server's desire to authenticate **me**.

**Comment by jeroen** [comment]
We [switched to schannel](https://ropensci.org/technotes/2017/10/10/curl-30/) in curl 3.0 so that is probably why you weren't seeing this problem before.

**Comment by rlmoore0** [comment]
Per the recommendation, I set ssl_verifypeer to FALSE.  I also deleted the line that set cainfo, just to be sure.  

curl 4.0 returns 

Error in curl::curl_fetch_memory("https://jira.ud.tools.gov/jira/rest/auth/1/session", schannel: Failed to get certificate location for
 U :/private/Certificates/rlrnoorl .ds_complete. pem

curl 2.8.1 returns code 200 with appropriate data.

The certificate file that the system is complaining about is the cert about my identity, which the server is using to confirm that I'm a legitimate user.

I also tried this with httr.  (I know httr depends on curl.)  I swapped curl 4.0 and 2.8.1, just to see if I got any other error message.  (Setting libcurl parameters with httr::set_config and httr::config.)  Same error messages with curl 4.0 installed; success with curl 2.8.1.

**Comment by jeroen** [comment]
So where is it picking up on that path `U :/private/Certificates/rlrnoorl .ds_complete. pem`? You must be setting that somewhere?

Can you add `verbose = TRUE` to your `handle_options()` so we can see what is going on?

**Comment by rlmoore0** [comment]
Two pieces to this:

**#1 Maybe this answer gives a hint? https://stackoverflow.com/a/55388090/318752**

Yes, I had tried that before I opened this ticket.  The problem is that I can't put my certificate into the MS cert store (schannel) because doing so is forbidden by government system policy.  The government has its own cert store (CSPID) which curl doesn't know about.  When using OpenSSL, the system lets me point to my certificate, which avoids this issue.  When using WinSSL, the system insists on the MS cert store, which I can't use.  And the use of schannel is hard-coded into curl 4.0, I understand?
 
**2 Where is that cert coming from?**

It is picking up my personal cert from 

sslcert = "U:/private/Certificates/rlmoor1.ds_complete.pem"

(In libcurl, that's CURLOPT_SSLCERT.)  That's the ssl client certificate, not the server/host cert chain that is being set through cainfo and ssl_verifypeer.

With verbose = TRUE:

•Hostname in DNS cache was stale, zapped
•Trying 8.175.43.168...
•TCP NODELAY set
•Connected to jira.ud.tools.gov (8.175.43.168) port 443 (#1)
•schannel: Failed to get certificate location for U :/private/Certificates/rlmoorl.dscomplete.pem
•Closing connection 1
•schannel: shutting down SSL/TLS connection with jira.ud.tools.gov port 443
 Error in curl::curl_fetch_memory("https://jira.ud.tools.gov/jira/rest/auth/1/session", schannel: Failed to get certificate location for
 U :/private/Certificates/rlrnoorl .ds_complete. pem

**Comment by jeroen** [comment]
OK so that is tricky. The current CRAN version indeed hardcodes the use of schannel on Windows, so you are expected to stick your certs in your MS store.

To work around this, I have just pushed a fix to the master branch that changes the behavior of curl to respect the `CURL_SSL_BACKEND` variable. Can you try this:;

```r
install.packages("https://github.com/jeroen/curl/archive/master.tar.gz", repos = NULL)
```

Now if you set `CURL_SSL_BACKEND="openssl"` before starting R, you should be able to use curl with the openssl backend, which then supports `cainfo` and `sslcert` options like in 2.8.1.

**Comment by rlmoore0** [comment]
> set CURL_SSL_BACKEND="openssl" before starting R

I presume you mean as a Windows environment variable in the OS?

**Comment by jeroen** [comment]
You can also set it in your user `.Renviron` file, see the manual page for `?Startup` in R.

**Comment by rlmoore0** [comment]
I added CURL_SSL_BACKEND="openssl" in my Rprofile.SITE file.  
I did remove.packages("curl") and, for good measure, remove.packages("httr").  I closed & restarted R.
I then reinstalled from the variant branch.  (httr was reinstalled from cran after installing the curl variant.)

I exited R.  I restarted R.  I confirmed that CURL_SSL_BACKEND was set:

> CURL_SSL_BACKEND
> [1] "openssl"

It gave me the same error message:

> Error in curl::curl_fetch_memory(url, handle = handle) :
>   schannel: Failed to get certificate location for U:/private/Certificates/rlmoor1.ds_complete.pem
